### PR TITLE
Charts

### PR DESCRIPTION
**Pull request by martin-ebay**
- [x] This PR contains CSS changes
- [ ] This PR does not contain CSS changes

## Description
adding a handful of tokens for light and dark for data viz colors

added the donut-chart and chart-legend including documentation.

## Notes

- [x] I verify the build is in a non-broken state
- [x] I verify all changes are within scope of the linked issue

- [x] I regenerated all CSS files under dist folder
- [x] I tested the UI in all supported browsers
- [ ] I did a visual regression check of the components impacted by doing a Percy build and approved the build
- [x] I tested the UI in dark mode and RTL mode
- [x] I added/updated/removed Storybook coverage as appropriate

### CODE REVIEW

File: src/less/donut-graph/donut-graph.less
Comment (ArtBlue) [inline src/less/donut-graph/donut-graph.less:21]: In this case, there may not be much of a performance boost, but optionally, you could add `will-change: opacity;` here to initialize browser optimization ahead of time for a potential transition.

**Review by ArtBlue — COMMENTED** [review]
Overall, great job, Martin! Just a few minor things, and one thing that is probably more notable - the printer issue.

File: docs/_includes/donut-graph.html
Comment (martin-ebay) [reply docs/_includes/donut-graph.html:9]: looks like this was depreciated according to https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/version.  I'll remove the version attributes

File: docs/_includes/chart-legend.html
Comment (ArtBlue) [inline docs/_includes/chart-legend.html:7]: It's fair to say it's implied, but it would be better to be explicit about this; in the paragraph above let's state that the chart legend will fit the width of its container.

**Review by ArtBlue — APPROVED** [review]
Looks good! Just add that bit of text.

**Comment by martin-ebay** [comment]
@ArtBlue updated

File: docs/_includes/chart.html
Comment (ianmcburnie) [inline docs/_includes/chart.html:7]: Do we need `chart--stacked`? For the default/base case we should typically have no modifiers. For variations we start adding modifiers.

File: docs/_includes/chart.html
Comment (ianmcburnie) [inline docs/_includes/chart.html:66]: Looks like the example code is out of sync with the real code above. E.g. `.chart--composition` class.

File: docs/_includes/chart.html
Comment (ianmcburnie) [inline docs/_includes/chart.html:8]: So this tabindex=0 on the root means a users has to make an additional TAB key press to get into the widget. I think we can remove it so that tabbing lands us directly onto the first slice (similar kb behaviour to a radio button group).

File: docs/_includes/chart.html
Comment (martin-ebay) [reply docs/_includes/chart.html:7]: I was trying to avoid overwriting any properties from the base chart class, so chart--stacked was needed for a couple of additional properties.  I could move them to the base chart class but will have to override a couple properties for the other layouts

Comment (martin-ebay) [reply docs/_includes/chart.html:7]: for more clarity the chart--stacked is setting a max width, which would bi overwritten for the chart--composition and chart--horizontal version.  Let me know if you would prefer I remove the --stacked class and just override that max-width

File: docs/_includes/chart.html
Comment (martin-ebay) [reply docs/_includes/chart.html:66]: good catch, missed this during some cleanup

File: docs/_includes/chart.html
Comment (martin-ebay) [reply docs/_includes/chart.html:8]: 👍

**Comment by ianmcburnie** [comment]
One minor nag, the indentation in the code examples is off, if you compare with the others.

File: docs/_includes/chart.html
Comment (ianmcburnie) [inline docs/_includes/chart.html:9]: I don't think that `graphics-document` and `graphics-symbol` are ARIA roles. If not, these should be removed.

File: docs/_includes/chart.html
Comment (ianmcburnie) [inline docs/_includes/chart.html:10]: Rather than `<desc>`, should we be leveraging  `<figure>` and `<figcaption>`?

File: docs/_includes/chart.html
Comment (ianmcburnie) [inline docs/_includes/chart.html:40]: Do we need `chart-legend__container-vertical` if vertical is the default? Also, syntax should be a double-dash, not single dash.

File: docs/_includes/chart.html
Comment (martin-ebay) [reply docs/_includes/chart.html]: 👍

File: docs/_includes/chart.html
Comment (martin-ebay) [reply docs/_includes/chart.html]: removing it

**Comment by ianmcburnie** [comment]
@martin-ebay Seems like we are missing the graph title?

Also, if we go with `<figure>` perhaps all of this "metrics" forms the `<figcaption>`?

Figma:

Screen Shot 2022-11-30 at 4 02 32 PM

Dev:

Screen Shot 2022-11-30 at 4 12 16 PM

**Comment by martin-ebay** [comment]
@ianmcburnie in regards to `Also, if we go with <figure> perhaps all of this "metrics" forms the <figcaption>?` The metrics are an optional display so I don't think we would want to link the figure/figcaption to it.  

Also in regards to the missing title, I was avoiding having it baked into the component as instead there is a parameter int component for a aria-labeledby for the title.

**Comment by martin-ebay** [comment]
@ianmcburnie also on the figcaption highcharts does not use this anywhere, so I'm not sure if we want to move our description into it.  With the svg <desc> tag it is caught and read out by the screen reader.

**Comment by ianmcburnie** [comment]
> @ianmcburnie also on the figcaption highcharts does not use this anywhere, so I'm not sure if we want to move our description into it. With the svg tag it is caught and read out by the screen reader.

With Safari & Voiceover, the `<desc>` does not get announced when kb focus moves inside of the widget. This is expected behavior. If we were to use figure & figcpation, the figcaption would be read out when kb focus moves into the figure. Same goes for aria-label/aria-labelledby on the root if we were to give the root a role of "group".

> Also in regards to the missing title, I was avoiding having it baked into the component as instead there is a parameter int component for a aria-labeledby for the title. This was based on some older feedback. Originally I had the title as part of the svg title tag

But the design is showing a _visible_ title. I'm under the impression that the words "Graph title" would be replaced with more realistic content, like "Browser market share"? So that's nothing to do with aria.

**Comment by ianmcburnie** [comment]
@martin-ebay Hold off on any changes for a bit, I just need to make a few small changes. Will probably save us on back and forth.

**Comment by ianmcburnie** [comment]
@martin-ebay I notice donut is called "donut-chart" in some places and "donut-graph" in others. Let's change all references to be donut-chart.

**Comment by martin-ebay** [comment]
@ianmcburnie they are different because I you asked me to break them apart at one point.  So I have a donut graph that is just the svg portion and then donut-chart is the composed version with the layout options

File: docs/_includes/chart.html
Comment (ianmcburnie) [reply docs/_includes/chart.html]: I removed these.

**Comment by ianmcburnie** [comment]
> @ianmcburnie they are different because I you asked me to break them apart at one point. So I have a donut graph that is just the svg portion and then donut-chart is the composed version with the layout options

I see. The module named "chart" is now the composition piece with the layout options. If we are going with `donut-graph` for the standalone SVG, I think there are still one or two references that need updating.

**Comment by martin-ebay** [comment]
@ianmcburnie are there particular areas you see that need updated?  I'm not finding anything that looks out of place between the `donut-graph` and `chart` classes

**Comment by ArtBlue** [comment]
> @ianmcburnie are there particular areas you see that need updated? I found a couple references to donut-chart in the variable names for the colors of the graph, and looks like you may have added some in the docs that I changed to chart for now as I updated the donut-chart to chart in the skin/docs/index.html

I'm actually seeing some issues too that I hadn't noticed before. I don't want to deviate too much from the direction Ian wanted to go and I don't know previous convos on/off-line, but here are some notes:

1. Isn't `chart` too broad? Shouldn't we simply use `donut-chart`? Or if the donut chart is a modifier of `chart`, maybe it should be `.chart .chart--donut`? 
2. The organization of `chart` with a `chart--composition` modifier is a bit awkward IMO. 
image
3. Shouldn't there be a `chart--small` and `chart--large` modifier on `chart`?
4. This is a bit odd: `donut-graph chart-graph__container chart-graph__container--composition`. I don't see any `.chart-graph` above so `chart-graph__container` on `donut-graph` is a bit confusing.

I stopped looking after a bit because it might be good to get on the same page as far as structuring things.

I think it would help immensely if we can first agree on an information architecture of the component(s) - what they are called, what modifiers they have, and how they are used in conjunction with each other.

**Comment by ianmcburnie** [comment]
@ArtBlue 

Even though I knew they were needed, I didn't make any major markup or css changes in my commit. I mainly just put the information architecture in place:

1. donut-graph 
2. chart-legend
3. chart

I believe chart is useful as a standalone module so that it can take on the role of layout and composition of the various parts of a chart.

The markup for 1 and 2 looks in pretty good shape. However 3 needs re-working. I'd be tempted to strip it right back to the minimum structure and see what is really needed. Both small and large HTML structures should begin with the figcaption -  I noticed that large isn't doing this - so maybe start with that change and build up from there.

**Comment by martin-ebay** [comment]
@ianmcburnie I know you mentioned using <figcaption>, I was just reviewing the highcharts implementations and they all use the <figure> tag as a wrapper, then they still use the <desc> tag inside of the svg.

Do we want to deviate from that pattern for our one custom chart?

Also are we planning for additional charts in the future?  I'm asking as the generic `chart` class / component might be confusing if it is only used for the single chart component.  Even the way we have the documentation for the `chart` could be confusing when it only applies to the donut-chart and none of the other chart components utilize it.

Highcharts doesn't let us break out the legend from the single svg it generates so we have no other use cases for the `chart` component.

I'm thinking maybe we should do a little bit of a comparison to all the highchart chart implementations that only utilize the new token values from skin.

**Comment by ianmcburnie** [comment]
> Do we want to deviate from that pattern for our one custom chart?

Now I think about it, for the standalone `donut-graph` we don't actually need the figure element. We can just use a div and put the `<desc>` back.  For the full chart, I do like `<figcaption>` as the screen reader will announce this caption as soon as focus moves inside of the figure. The svg `<desc>` will only be read out when focus moves explicitly onto the svg tag itself. We could keep `<desc>` also in that case, but I think it's going to feel redundant if the caption has already taken care of describing things.

> Even the way we have the documentation for the chart could be confusing when it only applies to the donut-chart and none of the other chart components utilize it.

Direct users of skin should not be confused by this, as there are no other charts currently in skin. For users of coreui, it's a black box, i.e. they don't know/care whether it's HighCharts or a Skin chart.

The chart is a composite class for laying out a graph, legend and caption. If we add more graphs in future, rather than duplicating layout css and logic in two different graph modules, the chart module handles it all in one place. It's very much like the field module.

**Comment by ArtBlue** [comment]
@ianmcburnie , while I realize there will at first be only one type of chart, I'm struggling a bit with `chart` standing in for the `donut-chart`. I think this makes scaling much more difficult if we end up adding more simple chart types (eg. bar chart) in the future. I think this would make things more scalable:

`````
chart
|  |
|  chart-legend
|  |
|  donut chart
|  |  |
|  |  donut-graph
|  | 
`````

I'm still working through this in my head, so it's still muddy. Maybe `chart-donut` can be a modifier on `chart`? Dunno.

IMO, if we continually try to move the implementation more and more towards the Highcharts implementation, we should really start to question whether we really need this in Skin to begin with.

**Comment by ianmcburnie** [comment]
> IMO, if we continually try to move the implementation more and more towards the Highcharts implementation, we should really start to question whether we really need this in Skin to begin with.

Great point! 

>  I'm struggling a bit with chart standing in for the donut-chart

Chart doesn't stand in for donut chart. Chart is purely a *layout* module. Think of it as 'chart-layout'. The idea is that the 'chart-layout' might offer different layouts and combinations of graphs, metrics and legends so that this could all be handled in one central place...

That being said, the more I look through all the different types of charts and graphs, the two layouts that we have for a donut-chart (small and large) do feel very specific to donut-chart and perhaps not usable by any other chart or graph? So the more I look at it I'm not sure how much mileage we get out of that modularity. Tempted to just flatten everything back into a single, self-contained 'donut-chart' module to unblock us, as the more pressing priority is figuring out the responsiveness.

**Comment by ianmcburnie** [comment]
... the BEM is going to be really strange though, once we get down into the guts of it. For example, the legend items. BEM's naming scheme isn't really well suited once we have that many of levels. We'd have `donut-chart__legend` ... but we can't do `donut-chart__legend__item`. This is what we ran into the first time we attempted it. This is a hint that things should be separated out into standalone blocks (e.g. `chart-legend` and `chart-legend__item`).

**Comment by martin-ebay** [comment]
What if we leave chart-legend broken out and combine the chart / donut-chart.  Or we could even make the chart legend a little more generic and name it legend

**Comment by martin-ebay** [comment]
also I think we could make the responsiveness work with flex by switching between row and columns and setting the order on the elements as needed.

**Comment by martin-ebay** [comment]
@ianmcburnie when you mention an 'inline' version of the legend are you talking about the ones for some of the other charts?  If so we won't be able to use this chart legend for those components as highcharts bakes the legends into their single svg container implementation.

If we want to move away from highcharts we could re-use it with other charts.

**Comment by ianmcburnie** [comment]
> @ianmcburnie when you mention an 'inline' version of the legend are you talking about the ones for some of the other charts? If so we won't be able to use this chart legend for those components as highcharts bakes the legends into their single svg container implementation.
> 
> If we want to move away from highcharts we could re-use it with other charts.

Yeah, something like this: 

Screen Shot 2022-12-08 at 12 15 30 PM

We aren't moving away from HighCharts, but an inline layout of a legend might in itself be useful in a CSS library such as Skin... if we were looking at making `chart-legend` more generic.

**Comment by martin-ebay** [comment]
Is it worth building the versions that we have no use case for now, or could we add it as an enhancement later?

**Comment by ianmcburnie** [comment]
> Is it worth building the versions that we have no use case for now, or could we add it as an enhancement later?

I don't think it's worth building now. Seems like there is a good forward path to adding it when/if we ever need.

**Comment by ianmcburnie** [comment]
For the standalone donut-graph we should probably show an example with tooltips.

Screen Shot 2022-12-08 at 3 27 06 PM

**Comment by martin-ebay** [comment]
for showing it with tooltips should I just show one hard coded into a spot, or should I some how implement the JS to make tooltips work in skin.  It involves tracking the mouse position and adjust contents based on the hovered index of the graph segment.  It doesn't just attach a tooltip to a host object like most of the other components.

**Comment by ianmcburnie** [comment]
> for showing it with tooltips should I just show one hard coded into a spot, or should I some how implement the JS to make tooltips work in skin. It involves tracking the mouse position and adjust contents based on the hovered index of the graph segment. It doesn't just attach a tooltip to a host object like most of the other components.

We mainly just need to see the DOM order & HTML structure and how that might change or be affected with the JS layer.

**Pull request closed without merging**
